File: internal/domain/attachment.go
package domain

import (
	"errors"
	"time"

	"github.com/RobinThrift/belt/internal/auth"
)

var ErrCreateAttachment = errors.New("error creating attachment")
var ErrAttachmentNotFound = errors.New("attachment not found")

type AttachmentID int64

type Attachment struct {
	ID               AttachmentID
	Filepath         string
	OriginalFilename string
	ContentType      string
	SizeBytes        int64
	Sha256           []byte
	CreatedBy        auth.AccountID
	CreatedAt        time.Time
}

type AttachmentList struct {
	Items []*Attachment
	Next  *string
}

File: app/src/domain/Attachment.ts
import type { Temporal } from "temporal-polyfill"

import { encodeToBase64 } from "@/lib/base64"

export type AttachmentID = string

export interface Attachment {
    id: AttachmentID

    filepath: string
    originalFilename: string
    contentType: string
    sizeBytes: number
    sha256: Uint8Array<ArrayBuffer>
    createdAt: Temporal.ZonedDateTime
}

export interface AttachmentList {
    Items: Attachment[]
    next?: string
}

export const ATTACHMENT_BASE_DIR = "attachments"

export function parseAttachmentURL(
    src: string,
): { attachmentID: string; metadata: Record<string, string> } | undefined {
    let u: URL
    try {
        u = new URL(src)
    } catch {
        return
    }

    if (u.protocol !== "attachment:") {
        return
    }

    return {
        attachmentID: u.hostname || u.pathname,
        metadata: Object.fromEntries(u.searchParams.entries()),
    }
}

export function imageBlobToSrc({
    data,
    mime,
}: {
    data: Uint8Array<ArrayBuffer>
    mime: string
}): string {
    if (mime === "image/svg+xml") {
        return `data:image/svg+xml;charset=utf-8;base64,${encodeToBase64(data)}`
    }

    return URL.createObjectURL(new Blob([data]))
}
